### ISSUE TITLE
Separate development, build and test dependencies from runtime dependencies

### ISSUE THREAD

**Issue by Korijn** [conda/conda-env#200]
**Introduction**
Similar to how `npm`'s `package.json` has a `dependencies` and `devDependencies` node, it would be interesting to make such a distinction in `environment.yaml`.

This would be useful to have tools such as `pyinvoke`, `flake8` and `sphinx` listed in `devDependencies`, since they are not required for an environment where you will only run the package code (for example, when deploying).

Also see the dicussion here: http://stackoverflow.com/questions/18875674/whats-the-difference-between-dependencies-devdependencies-and-peerdependencies

On a related note, I use some command line automation (accomplished with `pyinvoke`) to keep my `environment.yaml` and my conda recipe's `meta.yaml`'s requirement lists (run/build/test) in sync, but since we can't separate development dependencies from runtime dependencies yet, that complicates things. This inconvenience could also be solved with this feature.

**Plan**
Here's the plan after a few iterations and some discussion:

The "root" runtime dependencies list will correspond to the `requirements\run` list in `meta.yaml`. Other sublists extend the runtime dependencies (see https://github.com/conda/conda-env/issues/200#issuecomment-155728552) to avoid duplication of items. That would lead to the following to-do list for this feature:

- [ ] parse subsections in the `dependencies` node (e.g., for `dev`elopment, `test`ing and `build`ing)
- [ ] add optional argument `-e subsection` to CLI to install dependencies from the root plus the specified subsection; install everything if the optional argument is left out.
- [ ] then there are two options to install only root dependencies:
    - add optional argument `--production` to CLI. This obviously cannot be used in conjunction with the `-e` parameter.
    - specify `-e run` or `-e production` (any suggestions?).

Example of such a file:

```yaml
name: myproject
dependencies:
  - python
  - boltons
  - dev:
    - flake8
    - sphinx
    - pip:
      - invoke
  - test:
    - pytest
    - pytest-mock
    - pytest-flakes
  - build:
    - boost
    - gcc
```

The big benefit of this format is that you maintain backwards compatibility (in the sense that existing `environment.yaml` files don't need to change to continue to work as expected).

There is one additional work item that should be completed in order to let `environment.yaml` grab its contents from `meta.yaml` dynamically, so that the requirements lists don't need to be duplicated across both files. De-duplication of the `meta.yaml` lists can be handled there as well.

- [ ] merge PR #88 and add pyyaml to the available modules in the templates so that `environment.yaml` can retrieve its contents from `meta.yaml`

And as noted by @nicoddemus in https://github.com/conda/conda-env/issues/200#issuecomment-155736887, perhaps aliasing `dependencies` as `requirements` will improve user experience when working with `meta.yaml` and `environment.yaml` files.

- [ ] `dependencies` in `environment.yaml` gets an alias `requirements`

**Comment by malev** [comment]
Thanks for your suggestion @Korijn, I like your idea

**Comment by Korijn** [comment]
Any thoughts on what modules would be affected by such a change? I'm very eager to implement it soon... :)

**Comment by nicoddemus** [comment]
This is an interesting idea, and we have been thinking about how environments and their dependencies play out during the various stages of development.

Besides production, there might be other types of environments with different dependencies, for example running tests vs. production vs. compilation of C++ extensions. I think it makes sense to not tie different environments to "production" and "development" only, but to take a more general route of supporting multiple environment "types". [tox][tox] follows a similar approach where each test env can have it own set of dependencies.

[tox]: https://testrun.org/tox/latest/

Here's a draft of what multiple environment types would look like in an `environment.yml` file:

```yml
name: myproject

dependencies:
	- python
    - boltons

dependencies[dev]:
	- pyinvoke
    - flakes
    - sphinx
```

Entries in the standard `"dependencies"` are automatically inherited by all others, so to create a development environment you would call: `conda env create -e dev` (borrowing the `-e` syntax from tox).

Here's another example which shows multiple environments more close to our needs at work:

```yml
name: myproject

dependencies:
	- python
    - boltons

dependencies[tests]:
	- pytest
    - pytest-mock
    - pytest-flakes

dependencies[compile]:
	- boost
    - gcc
```

`conda env update` would also need to understand the `-e` option, so the user can quickly switch between different environments as neede.

**Comment by Korijn** [comment]
Another thing to keep in mind is that you need to somehow synchronize these dependency specifications with the conda recipe's `meta.yaml`. I'd really enjoy being able to refer to `environment.yaml` from my `meta.yaml`.

That would mean that you need at least the following three requirement lists to match [the recipe lists](http://conda.pydata.org/docs/building/meta-yaml.html#requirements-section), which matches @nicoddemus examples:

- run
- build
- test

Are there any other things you would ever want to list, anyway?

On another note, I would suggest that installing everything as default, and limiting to specific lists using arguments, is a more sane default for the CLI.

**Comment by nicoddemus** [comment]
> Another thing to keep in mind is that you need to somehow synchronize these dependency specifications with the conda recipe's meta.yaml.

Hmmm that's a problem we don't have as we always develop in `develop` mode and do not create packages for our internal projects, but I certainly see your point. We have contributed a PR in #88 which adds `jinja2` support and which we have been using extensively, perhaps it can be used to read `meta.yml` and fill your `dependencies.yml` section, even if it is a little hackish.

> Are there any other things you would ever want to list, anyway?

I'm not sure, probably not. Using the same possible requirement lists as conda-build certainly makes sense.

> On another note, I would suggest that installing everything as default, and limiting to specific lists using arguments, is a more sane default for the CLI.

Hmmm agreed.

In the light of this discussion then, I think the `environment.yml` format should match conda-build's:

```yml
name: myproject
dependencies:
  run:
    - python
    - boltons
  test:
    - python
    - boltons
    - pytest
```

A problem with this approach is that you have to duplicate the entries that appear in more than one requirement list... for us this would be a little unwieldy since our internal projects easily depend on dozens of packages. 

(Just realized that is a shame that conda-env uses `dependencies` and conda-build uses `requirements`)

**Comment by Korijn** [comment]
The jinja2 feature is interesting! That would definitely work, if you could parse the `meta.yaml` file in the template. Is it possible to have access to pyyaml in the template?

> (Just realized that is a shame that conda-env uses `dependencies` and conda-build uses `requirements`)

You could just allow newer versions of conda-env to use `requirements` as well, as an alias.

**Comment by nicoddemus** [comment]
>  Is it possible to have access to pyyaml in the template?

Not currently, but that should be simple to add.

> You could just allow newer versions of conda-env to use requirements as well, as an alias.

Yep, not really a problem. :smile:

**Comment by Korijn** [comment]
You could just leave the `run` section as the root, so that the other subsections "extend" the runtime dependencies, to avoid duplicating everything. That would lead to the following:

- [ ] `dependencies` in `environment.yaml` gets an alias `requirements`
- [ ] parse subsections in the `dependencies` node for `dev`elopment, `test`ing and `build`ing
- [ ] add optional argument `--production` to CLI to only install runtime dependencies; install everything if the optional argument is left out

Example of such a file:

```yaml
name: myproject
requirements:  # alternatively, `dependencies:` would also be OK
  - python
  - boltons
  dev:
    - flake8
    - sphinx
  test:
    - pytest
    - pytest-mock
    - pytest-flakes
  build:
    - boost
    - gcc
```

The big benefit of this format is that you maintain backwards compatibility.

The only problem with this is that you can't directly synchronize the `meta.yaml` requirements lists into `environment.yaml` with the jinja2 feature, since those lists still contain duplicate entries.

**Comment by Korijn** [comment]
I went ahead and updated the issue description and title to reflect the current state of the discussion.

**Comment by Korijn** [comment]
Another potential issue I stumbled upon today is that we can't list pip requirements in `meta.yaml`. :/

**Comment by nicoddemus** [comment]
True. There's a new format being discussed in #145.

**Comment by malev** [comment]
@Korijn `conda-env` and `conda-build` are different projects and I don't think there is any plan of using the `environment.yml` file from `conda-build`. At least for now. Although it think it would be amazing.
I like
``` yaml
name: myproject

dependencies:
    - python
    - boltons

dependencies[tests]:
    - pytest
    - pytest-mock
    - pytest-flakes

dependencies[compile]:
    - boost
    - gcc
```
We can have it like: whatever is in dependencies is common and then you can be more specific depending on the task that you want to do (tests, build or compile). To be honest I don't like the `-e` flag. It feels like `environment` so we are using conda **environment** with an **environment** flag and that feels weird. In that scenario I personally like the [bundler approach](http://bundler.io/v1.10/groups.html).
@nicoddemus if we implement this do you think we won't need `jinja2` anymore? or you think it's still useful?

**Comment by Korijn** [comment]
Sounds like a worthwhile improvement to me! Although I'm a bit dissapointed that I'll have to continue to resort to command line automation to keep my requirements synchronized between environment.yaml and meta.yaml.

I like the bundler approach, the `--without` flag seems useful. I'd like to see a `--with` flag as an alternative option as well though. There are cases where you just want to exclude everything except your `--with` selection, so as to avoid the risk that your `--without` exclusion list is non-exhaustive.

**Comment by nicoddemus** [comment]
> I like the bundler approach, the --without flag seems useful. I'd like to see a --with flag as an alternative option as well though.

I agree, `--with` fits our use case nicely. Here's what our continuous integration build script for an hypothetical library named `myproject` would look like:

```
cd myproject

conda env create --force --with=compile
source activate myproject
python setup.py develop  # or: conda develop src/python
scons

conda env update --prune --with=tests
py.test
```

Of course, here `--with=tests` means `dependencies` + `dependencies[tests]`. 

Does this make sense to you guys?

> @nicoddemus if we implement this do you think we won't need jinja2 anymore? or you think it's still useful?

One use case where we still use `jinja2` is to have different dependencies based on the platform, like this:

```yml
name: myproject

dependencies:
    - python
    - boltons
    {% if sys.platform.startswith('win') %}
    - pywin32
    {% endif %}
```

And we use that extensively in our projects. 

We could easily live with `conda-build`'s approach of using a special pre-processed comment instead:

```yml
name: myproject

dependencies:
    - python
    - boltons
    - pywin32  # [win]
```

One benefit of the pre-processed comments is that they are valid `yaml` files and can be processed by standard tools.

But `jinja2` support / pre-processed comment is off-topic, of course. :smile:

**Comment by Korijn** [comment]
So how would one specify to install only the default deps? `conda env create --without=all` and/or `conda env create --with=default`?

And +1 on the preprocessed comments suggestion.

**Comment by nicoddemus** [comment]
> I guess it would make more sense to use --with=all and install only the default deps without both flags?

I think so too, at least for the use cases I have in mind. So, to summarize:

**Proposal 1**

```yml
name: myproject

dependencies:
    - python
    - boltons

dependencies[tests]:
    - pytest
    - pytest-mock
    - pytest-flakes

dependencies[compile]:
    - boost
    - gcc

dependencies[deploy]:
    - gunicorn
```

* `conda env create`: uses `dependencies`;
* `conda env create --with=tests`: uses `dependencies` + `dependencies[tests]`;
* `conda env create --with=tests,compile`: uses `dependencies` + `dependencies[tests]` + `dependencies[compile]`;
* `conda env create --with=all`: uses `dependencies` + `dependencies[tests]` + `dependencies[compile]` + `dependencies[deploy]`;

I can't think of a use case for using a `--without` option though.

What you guys think? Once we agree on the format and behavior, I can tackle this if nobody else wants to. :smile:

**Comment by Korijn** [comment]
I like it! How about just "groups"?

**Comment by nicoddemus** [comment]
:+1: for groups :smile:

**Comment by Korijn** [comment]
There are cases where you want `conda env create` to just install everything, or a default selection, for example, when you don't have control over the command that is being executed (e.g., readthedocs is implementing conda environments by just calling `conda env create`).

But the semantics of the command line options will get confusing if the omission of `--with` leads to everything being installed. On the other hand, if you'd use `--without`, there is a risk that your exclusion lists are not exhaustive.

I think that in order to get around this, there would need to be an additional key in the `environment.yaml` file that specifies the default groups to be installed. If the key is omitted, nothing will change about the behaviour that @nicoddemus specified above. So, considering this file:

```
name: myproject

dependencies:
    - python
    - boltons

dependencies[tests]:
    - pytest
    - pytest-mock
    - pytest-flakes

dependencies[compile]:
    - boost
    - gcc

dependencies[deploy]:
    - gunicorn
```

- `conda env create`: uses `dependencies`;
- `conda env create --with=tests` uses `dependencies + dependencies[tests]`;
- `conda env create --with=tests,compile` uses `dependencies + dependencies[tests] + dependencies[compile]`;
- `conda env create --with=all` uses `dependencies + dependencies[tests] + dependencies[compile] + dependencies[deploy]`;

If we update the file by adding the key `default`, only the default behaviour will change:

```
name: myproject

default:
    - tests
    - compile

dependencies:
    - python
    - boltons

dependencies[tests]:
    - pytest
    - pytest-mock
    - pytest-flakes

dependencies[compile]:
    - boost
    - gcc

dependencies[deploy]:
    - gunicorn
```

- `conda env create`: uses `dependencies + dependencies[tests] + dependencies[compile]`;
- `conda env create --with=tests` uses `dependencies + dependencies[tests]`;
- `conda env create --with=tests,compile` uses `dependencies + dependencies[tests] + dependencies[compile]`;
- `conda env create --with=all` uses `dependencies + dependencies[tests] + dependencies[compile] + dependencies[deploy]`;

This will allow you to exercise control over default behaviour, for example to make sure that devs always get a full environment if they just call `conda env create` and to ensure that external systems such as readthedocs set up the environment appropriately.

Lastly, I think this is also an important use case to support:

```
name: myproject

dependencies:
    - all

dependencies[tests]:
    - pytest
    - pytest-mock
    - pytest-flakes

dependencies[compile]:
    - boost
    - gcc

dependencies[deploy]:
    - gunicorn
```

- `conda env create`: uses `dependencies + dependencies[tests] + dependencies[compile] + dependencies[deploy]`;
- `conda env create --with=tests` uses `dependencies + dependencies[tests]`;
- `conda env create --with=tests,compile` uses `dependencies + dependencies[tests] + dependencies[compile]`;
- `conda env create --with=all` uses `dependencies + dependencies[tests] + dependencies[compile] + dependencies[deploy]`;

**Comment by nicoddemus** [comment]
@Korijn I'm :+1: for having a "default group". Perhaps it would be clearer to name it `default-dependencies` or `default-groups` though?

While currently `environment.yml` is all about installing packages, I see the possibility of adding further features to it, like [environment variables](https://github.com/conda/conda-env/pull/88), so I personally would prefer to have a more specific name for the top-level entries.

**Comment by Korijn** [comment]
Seems reasonable! :+1:

**Comment by Korijn** [comment]
Or:

```
name: myproject

default:
    dependencies:
        - tests
        - compile

dependencies:
    - python
    - boltons

dependencies[tests]:
    - pytest
    - pytest-mock
    - pytest-flakes

dependencies[compile]:
    - boost
    - gcc

dependencies[deploy]:
    - gunicorn
```

**Comment by nicoddemus** [comment]
I personally prefer:

```yml
default-groups:
    - tests    
```

But really either way is fine to me. :grin:

**Comment by Korijn** [comment]
:+1: Fine by me!

I have no further objections, so unless @malev has any objections...

**Comment by nicoddemus** [comment]
Any news on this? We saw the need for something like this again this last week... :grin: n:

**Comment by Korijn** [comment]
I'm considering implementing preprocessed comments just like how they work in conda-build (so the usual suspects are available: `win`, `py27`, etc.), with the addition of being able to define more variables such as `test` yourself. I would opt for the `--with` parameter to set the custom variables to `True`. Special case: `--with all` will install everything.

```
name: myenv
dependencies:
  - python
  - nose  # [test]
  - sphinx  # [docs]
  - pywin32  # [win]
```

**Comment by Korijn** [comment]
See #213

**Comment by Korijn** [comment]
The PR is updated!

**Comment by Korijn** [comment]
Ping @ilanschnell and @kalefranz any chance this can be included in the release? :)

**Comment by nicoddemus** [comment]
Yes please. :smile:

**Comment by Korijn** [comment]
This is still GTG as far as I can tell. Can't it be merged in?

**Comment by Korijn** [comment]
See #213

**Comment by Korijn** [comment]
This was resolved by #213, closing.
